geo/src/algorithm/offset/line_intersection.rs

```diff
@@ -0,0 +1,321 @@
+/// > method in simplifyvw.rs. It is nice because it uses the orient2d method
+/// > of the Kernel, however it only gives a true/false answer and does not
+/// > return the intersection point or parameters needed.
+///
+/// We already have LineIntersection trait BUT we need a function that also
+/// returns the parameters for both lines described below. The LineIntersection
+/// trait uses some fancy unrolled code it seems unlikely it could be adapted
+/// for this purpose.
+///
+/// Returns the intersection point **and** parameters `t_ab` and `t_cd`
+/// described below
+///
+/// The intersection of segments can be expressed as a parametric equation
+/// where `t_ab` and `t_cd` are unknown scalars :
+///
+/// ```text
+/// a + ab · t_ab = c + cd · t_cd
+/// ```
+///
+/// > note: a real intersection can only happen when `0 <= t_ab <= 1` and
+/// >       `0 <= t_cd <= 1` but this function will find intersections anyway
+/// >       which may lay outside of the line segments
+///
+/// This can be rearranged as follows:
+///
+/// ```text
+/// ab · t_ab - cd · t_cd = c - a
+/// ```
+///
+/// Collecting the scalars `t_ab` and `-t_cd` into the column vector `T`,
+/// and by collecting the vectors `ab` and `cd` into matrix `M`:
+/// we get the matrix form:
+///
+/// ```text
+/// [ab_x  cd_x][ t_ab] = [ac_x]
+/// [ab_y  cd_y][-t_cd]   [ac_y]
+/// ```
+///
+/// or
+///
+/// ```text
+/// M·T=ac
+/// ```
+///
+/// Inverting the matrix `M` involves taking the reciprocal of the determinant
+/// (the determinant is same as the of the [cross_product()] of `ab` and `cd`)
+///
+/// ```text
+/// 1/(ab×cd)
+/// ```
+///
+/// Therefore if `ab×cd = 0` the determinant is undefined and the matrix cannot
+/// be inverted. The lines are either
+///   a) parallel or
+///   b) collinear
+///
+/// Pre-multiplying both sides by the inverted 2x2 matrix we get:
+///
+/// ```text
+/// [ t_ab] = 1/(ab×cd) · [ cd_y  -cd_x][ac_x]
+/// [-t_cd]               [-ab_y   ab_x][ac_y]
+/// ```
+///
+/// or
+///
+/// ```text
+/// T = M⁻¹·ac
+/// ```
+///
+/// Expands to:
+///
+/// ```text
+/// [ t_ab] = 1/(ab_x·cd_y - ab_y·cd_x)·[ cd_y·ac_x - cd_x·ac_y]
+/// [-t_cd]                             [-ab_y·ac_x + ab_x·ac_y]
+/// ```
+///
+/// Since it is tidier to write cross products, observe that the above is
+/// equivalent to:
+///
+/// ```text
+/// [t_ab] = [   ac×cd / ab×cd ]
+/// [t_cd] = [ - ab×ac / ab×cd ]
+/// ```
+
+fn line_segment_intersection_with_parameters<T>(
+    a: &Coord<T>,
+    b: &Coord<T>,
+    c: &Coord<T>,
+    d: &Coord<T>,
+) -> Option<(T, T, Coord<T>)>
+where
+    T: CoordFloat,
+{
+    let ab = *b - *a;
+    let cd = *d - *c;
+    let ac = *c - *a;
+
+    // TODO: I'm still confused about how to use Kernel / RobustKernel;
+    //       the following did not work. I need to read more code
+    //       from the rest of this repo to understand.
+    // if Kernel::orient2d(*a, *b, *d) == Orientation::Collinear {

```

Comment:
`Orientation` is an enum, so you have to use a [`match`](https://doc.rust-lang.org/book/ch06-02-match.html) for control flow.

something like

    match Kernel::orient2d(*a, *b, *c) {
        Orientation::Collinear => { // keep going }
        _ => { // not collinear, bail out or test different combination?? }
    }

As you note, the use of arbitrary "magic" threshold values without considerable justification and accompanying comments explaining the reasoning behind it for future maintainers is unlikely to be accepted (I've just finished removing an algorithm containing a lot of these that I had the misfortune of writing myself…), so if we can use existing mechanisms that would be much better.